### PR TITLE
GEODE-10276: Refactor PDX (de)serialization code

### PR DESCRIPTION
**Pull request by gaussianrecurrence**
- Code related to PDX de(serialization) has been refactored and aligned
   (in all possible means) to the Java client, solving the following
   issues:
    * GEODE-9968 - Fix deserialization for new fields in PdxSerializable
      class
    * GEODE-9753 - Coredump during PdxSerializable object serialization
    * GEODE-10220 - Coredump while initializing PdxType remoteToLocal
    * GEODE-10255 - PdxSerializable not working correctly for multiple
      versions of the same class
 - Reverted GEODE-8212, to solve potential issues during (de)serialization.
   This is due to the fact that the server orders PdxType fields
   sequentially, rather than alphabetically.
 - In order to achieve what was originally intended with GEODE-8212,
   now PdxInstances fields are written in alphabetical order, rather
   than in the order writeXX is called.
 - Also, code now should be easier to understand
 - Several tests modified to adapt to code changes
 - Added several UTs to further incrase coverage on PDX (de)serialization code.
 - Added several ITs to verify PDX (de)serialization scenarios.

### CODE REVIEW

**Comment by pivotal-jbarrett** [comment]
OMG! Awesome effort!

File: cppcache/src/CacheableObjectArray.cpp [hunk 1]
```diff
@@ -53,6 +54,18 @@ size_t CacheableObjectArray::objectSize() const {
...
+std::string CacheableObjectArray::getClassName() const {
+  if (!empty()) {
+    auto&& item = *begin();
+
+    if (auto pdx = dynamic_cast<PdxSerializable*>(item.get())) {
+      return pdx->getClassName();
+    }
+  }
+
+  return "java.lang.Object";
```
Comment (pivotal-jbarrett) [inline cppcache/src/CacheableObjectArray.cpp:67]: What if we make this string a `const` value in this compilation unit and change the return type of this method to return `const std::string&` so we aren't copy constructing strings on return.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]
I realize this is a draft but was excited to take a peek and noticed this one right away.

**Comment by gaussianrecurrence** [comment]
> I realize this is a draft but was excited to take a peek and noticed this one right away.

Yes, this is a WIP, there are some minor changes i yet want to make, and whole lot more of tests that I want to add, given the size of the change

File: tests/javaobject/PdxTests/DeliveryAddress.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+package PdxTests;
+
+import java.util.Vector;
+import org.apache.geode.pdx.PdxReader;
+import org.apache.geode.pdx.PdxSerializable;
+import org.apache.geode.pdx.PdxWriter;
+
+public class DeliveryAddress implements PdxSerializable
+{
+    String _addressLine;
```
Comment (pivotal-jbarrett) [inline tests/javaobject/PdxTests/DeliveryAddress.java:27]: Geode's Java conventions come from the Google Java Style Guide. Please don't prefix member variables. Also please make this private.

File: cppcache/integration/test/PdxInstanceTest.cpp [hunk 3]
```diff
@@ -268,7 +268,7 @@ TEST(PdxInstanceTest, testPdxInstance) {
       pdxTypeFromPdxTypeInstance->equals(*pdxTypeFromPdxTypeInstanceGet, false))
       << "PdxObjects should be equal.";
 
-  EXPECT_EQ(-960665662, pdxTypeInstance->hashcode())
+  EXPECT_EQ(514863279, pdxTypeInstance->hashcode())
```
Comment (pivotal-jbarrett) [inline cppcache/integration/test/PdxInstanceTest.cpp:271]: Why did this hash code change? Are we certain it matches the same hash that the server produces for it? If not this can lead to really bad single hop performance.

File: cppcache/integration/test/PdxSerializableTest.cpp [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  }
+
+  {
+    auto cache = cluster.createCache();
+    auto region = setupRegion(cache);
+    cache.getTypeRegistry().registerPdxType(
+        DeliveryAddress::createDeserializable);
+
+    auto entry = region->get("entry.v2");
+    EXPECT_TRUE(entry);
```
Comment (pivotal-jbarrett) [inline cppcache/integration/test/PdxSerializableTest.cpp:149]: This, and all the other `std::shared_ptr` nullability checks, should be `EXPECT_THAT(entry, IsNotNull())`.

File: cppcache/src/CacheImpl.hpp [hunk 5]
```diff
@@ -134,6 +134,12 @@ class APACHE_GEODE_EXPORT CacheImpl {
    */
   DistributedSystem& getDistributedSystem();
```
Comment (pivotal-jbarrett) [inline cppcache/src/CacheImpl.hpp:135]: Any chance this non `const` version can be refactored out?

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: cppcache/integration/test/PdxInstanceTest.cpp (hunk 3 above)
Comment (gaussianrecurrence) [reply cppcache/integration/test/PdxInstanceTest.cpp:271]: Damn, I did not thought of the fact that it was possible to use a PdxInstance as a key. Really thanks for pointing this out,

File: cppcache/src/CacheImpl.hpp (hunk 5 above)
Comment (gaussianrecurrence) [reply cppcache/src/CacheImpl.hpp:135]: Seems feasible :)

**Comment by gaussianrecurrence** [comment]
@pivotal-jbarrett I've addressed all of your comments, plus I've fixed some logic issues and introduced some changes, mostly to cover all edge cases. Additionally, added a bunch of new tests to verify the functionality refactored by the change. So long store sort, I've marked the PR as ready for review, any comments are welcomed :)

**Comment by gaussianrecurrence** [comment]
I haven't had the time to work on this for a long time or I think I'll have the opportunity to in the near future, hence I am closing this PR, if anyone wants to tackle this issue can take it up from its latest state.
